source/common/http/status.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <atomic>
+#include <string>
+
+#include "envoy/http/codes.h"
+
+#include "absl/base/attributes.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Http {
+
+/**
+ * Status codes for representing classes of codec errors.
+ */
+enum class StatusCode : int {
+  Ok = 0,
+  CodecProtocolError = 1,
+  BufferFloodError = 2,
+  PrematureResponse = 3,
+  CodecClientError = 4
+};
+
+/**
+ * Class for returning error status.
+ * This class is to replace the use of exceptions in Envoy's codecs.
+ * Requirements:
+ *   1. Represent the type system of existing codec exceptions.
+ *   2. Highly efficient when the status has NO error.
+ *   3. Register size, to avoid stack allocations.
+ *   4. Extensible.
+ *
+ * Constructing an error status is likely as heavy as throwing an exception due to a heap
+ * allocation.
+ */
+class ABSL_MUST_USE_RESULT Status final {
```

Comment:
Should we at least attempt to re-use abseil status and use the status payload for application-specific errors?
Having multiple statuses is a real pain.

Reply:
+1, I'd really like to understand what absl::Status is lacking and see if we can resolve that.

Reply:
Hm I see. We could potentially may our five (OK, CodecProtocol, BufferFlood, etc) to five of the absl canonical errors, and set up a kHttpCode payload to hold the `HttpCode` in the case of a `PrematureResponseException`. 

Other things we want are:
- a `StatusOr` -- maybe we need in-house `absl::variant` holding `absl::Status` that can be a drop in when absl::StatusOr is released?
- the potential to prevent overwriting a bad status (we catch and set a codec error, `ASSERT` we don't try to set it again).

Reply:
See my comment above also? Would all of this be solved via a template parameter/struct (which could be empty) added to the underlying status class?

Reply:
re: StatusOr -- it is about to be open sourced (probably weeks, not months). For now, the recommendation is to fork it (e.g. https://github.com/google/cel-cpp/tree/master/base), and let the abseil team know. Using abseil standard status should provide better experience with testing macros, proto/grpc status alignment, etc down the line I'd think.

Reply:
I've spent some time talking to the absl engineer about how to make absl::Status work with custom error spaces and here are the takeaways.

@ahedberg  The only question is how to reconsile absl::Status error space with the error space of Envoy dataplane.

- @mattklein123  It is highly unlikely that there will be ability to change error codes of absl::Status by any means. It is tied to the "canonical Google error space" and flexibility of changing it is outside of the scope of this class. This was actually considered before and discarded as it was adding unneeded (internally to Google) complexity.
- Recommended and the only way to modify error codes is to use absl::Status payloads which is a set of key/value pairs. Value is an opaque byte buffer. As suggested by @kyessenov 

So the API to the error facility might look something like this if we go with the absl::Status

```c++
namespace Envoy {
using Status = absl::Status;

enum class StatusCode : int {
  CodecProtocolError = 0,
  BufferFloodError = 1,
  PrematureResponseError = 2
};

// Methods for creating error codes
Status MakeCodecProtocolError(absl::string_view message);
Status MakeBufferFloodError(absl::string_view message);
Status MakePrematureResponseError(absl::string_view message, Http::Code http_code);

// Methods for checking and extracting error information
StatusCode GetStatusCode(Status);
bool IsCodecProtocolError(Status);
// ...

} // namespace Envoy
```

Internally these methods will create absl::Status with e.g. "Internal" error code and a payload with Envoy's error code and optionally the http_code. The extractors will get the error details from the payload. We could also do error code mapping as @asraa suggested. This would make allocation cost go up for one class of error only.

Downsides are:

1. Creating a payload needs two allocations
1. Extracting may need one too, unless we want to deal with the absl::Cord iterators and reconstruct payload byte by byte.
1. absl::Status can not be constructed directly and its `code()` method always return the same error code. It may be a gotcha for someone who does not know that there is additional API to be used for constructing and checking error codes.

The upside is that:
1. We do not need to maintain absl::Status. Although we will need to maintain the payload serializer/deserializer code.
1. absl::StatusOr comes for free as well as it will use absl::Status

Let me know your thoughts. It seems like people want to avoid forking these two classes and I agree this is not ideal.

Reply:
IMO the upsides of reusing the existing classes outweigh the negatives, especially because in the happy path there would be no extra allocations, so it seems fine to me. 

Conceptually though, I still don't understand why the payload mechanism couldn't be templated in such a way where the default implementation is the key/value pair map, and then we could have our own implementation with our own well specified struct. This would force StatusOr to also be templated on our templated Status type. This seems like the optimal solution, albeit a bunch more work, so I suppose it probably would not happen.